contracts/FeralfileArtworkV4.sol

```diff
@@ -76,37 +232,103 @@ contract FeralfileExhibitionV4 is
         return _contractURI;
     }
 
-    /// @notice isValidRequest validates a message by ecrecover to ensure
+    /// @notice verifies a message by ecrecover to ensure
     //          it is signed by owner of token.
     /// @param message_ - the raw message for signing
     /// @param owner_ - owner address of token
     /// @param r_ - part of signature for validating parameters integrity
     /// @param s_ - part of signature for validating parameters integrity
     /// @param v_ - part of signature for validating parameters integrity
-    function isValidRequest(
+    function verifySignature(
         bytes32 message_,
         address owner_,
         bytes32 r_,
         bytes32 s_,
         uint8 v_
     ) internal pure returns (bool) {
-        address signer = ECDSA.recover(
-            ECDSA.toEthSignedMessageHash(message_),
-            v_,
-            r_,
-            s_
+        return
+            ECDSA.recover(ECDSA.toEthSignedMessageHash(message_), v_, r_, s_) ==
+            owner_;
+    }
+
+    function _seriesExists(uint256 seriesId) internal view returns (bool) {
+        return _seriesMaxSupplies[seriesId] > 0;
+    }
+
+    function mintArtworks(
+        MintData[] memory data
+    ) external virtual onlyAuthorized {
+        for (uint256 i = 0; i < data.length; i++) {
+            _mintArtwork(data[i].seriesId, data[i].tokenId, data[i].minter);
+        }
+    }
+
+    function _mintArtwork(
+        uint256 seriesId,
+        uint256 tokenId,
+        address owner
+    ) internal {
+        // pre-condition checks
+        require(
+            _mintable,
+            "FeralfileExhibitionV4: contract doesn't allow to mint"
+        );
+        require(
+            owner != address(0),
+            "FeralfileExhibitionV4: mint to zero address"
+        );
+        require(
+            _seriesExists(seriesId),
+            string(
+                abi.encodePacked(
+                    "FeralfileExhibitionV4: seriesId doesn't exist: ",
+                    Strings.toString(seriesId)
+                )
+            )
         );
-        return signer == owner_;
+        require(!_exists(tokenId), "ERC721: token already minted");
+        require(
+            _seriesTotalSupplies[seriesId] < _seriesMaxSupplies[seriesId],
+            "FeralfileExhibitionV4: no slots available"
+        );
+
+        // mint
+        _totalSupply += 1;
+        _seriesTotalSupplies[seriesId] += 1;
+        _allArtworks[tokenId] = Artwork(seriesId, tokenId);
+        _safeMint(owner, tokenId);
```

Comment:
- with _safeMint, we need `IERC721Receiver` for this contract to receive tokens.